// @ogma/nestjs-module/src/ogma.constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OGMA_CONTEXT_PARSER = exports.OGMA_TRACE_METHOD_OPTION = exports.OGMA_INTERCEPTOR_PROVIDERS = exports.OGMA_REQUEST_SCOPED_SERVICE_TOKEN = exports.OGMA_SERVICE_TOKEN = exports.OGMA_INTERCEPTOR_SKIP = exports.OGMA_SERVICE_OPTIONS = exports.OGMA_INTERCEPTOR_OPTIONS = exports.MICROSERVICE_METADATA = exports.MESSAGE_METADATA = exports.OGMA_OPTIONS = exports.OGMA_CONTEXT = exports.OGMA_INSTANCE = void 0;
const core_1 = require("@nestjs/core");
const ogma_core_module_definition_1 = require("./ogma-core.module-definition");
exports.OGMA_INSTANCE = 'OGMA_INSTANCE';
exports.OGMA_CONTEXT = 'OGMA_CONTEXT';
exports.OGMA_OPTIONS = ogma_core_module_definition_1.MODULE_OPTIONS_TOKEN;
exports.MESSAGE_METADATA = 'message';
exports.MICROSERVICE_METADATA = 'microservices:pattern';
exports.OGMA_INTERCEPTOR_OPTIONS = 'OGMA_INTERCEPTOR_OPTIONS';
exports.OGMA_SERVICE_OPTIONS = 'OGMA_SERVICE_OPTIONS';
exports.OGMA_INTERCEPTOR_SKIP = 'OGMA_INTERCEPTOR_SKIP';
exports.OGMA_SERVICE_TOKEN = 'OGMA_SERVICE';
exports.OGMA_REQUEST_SCOPED_SERVICE_TOKEN = 'OGMA_REQUEST_SCOPED_SERVICE';
exports.OGMA_INTERCEPTOR_PROVIDERS = [exports.OGMA_INTERCEPTOR_OPTIONS, core_1.Reflector];
exports.OGMA_TRACE_METHOD_OPTION = 'OGMA_TRACE_METHOD_OPTION';
exports.OGMA_CONTEXT_PARSER = 'OGMA_CONTEXT_PARSER';
